HW2/P3/mandelbrot.pyx

```diff
@@ -41,9 +60,91 @@ cpdef mandelbrot(np.complex64_t [:, :] in_coords,
+            # Break each row into blocks of 8
+
+            # Need number of blocks on each row:
+            blox = in_coords.shape[1]/8
+
+            # Now I parallelize within each of those blocks.  
+            for block in range(blox):              
+                
+                # Take the initial real components of c.
+                creal = AVX.make_float8(inco_real[i, 8*block + 7],
+                                        inco_real[i, 8*block + 6],
+                                        inco_real[i, 8*block + 5],
+                                        inco_real[i, 8*block + 4],
+                                        inco_real[i, 8*block + 3],
+                                        inco_real[i, 8*block + 2],
+                                        inco_real[i, 8*block + 1],
+                                        inco_real[i, 8*block + 0])
+                
+                # Take the initial imaginary components of c.  
+                cimag = AVX.make_float8(inco_imag[i, 8*block + 7],
+                                        inco_imag[i, 8*block + 6],
+                                        inco_imag[i, 8*block + 5],
+                                        inco_imag[i, 8*block + 4],
+                                        inco_imag[i, 8*block + 3],
+                                        inco_imag[i, 8*block + 2],
+                                        inco_imag[i, 8*block + 1],
+                                        inco_imag[i, 8*block + 0])
+                
+                # Starting values of zero for z and the iteration counts.
+                zreal = avx0
+                zimag = avx0              
+                itercounts = avx0
+
+                for iter in range(max_iterations):
+                
+                    # Magnitude squared (sum of squares of real and imaginary components).
+                    magsquared = AVX.mul(AVX.mul(zreal,zreal),AVX.mul(zimag,zimag))
+                     
+                    # Determine whether the magnitude squared is less than 4
+                    mask = AVX.less_than(magsquared, avx4)
+                    
+                    # If all elements are less than zero, write the iteration counts to the array.
+                    # Also, if still iterating at the last iteration, write the maximum iteration count
+                    #    to the array.  And break out of the loop.  
+                    if AVX.signs(mask)==0 or iter==(max_iterations - 1):
+                        counts_to_output(itercounts, out_counts, i, block)
+                        break
 
+                    # If the loop hasn't hit break, convert the mask to 1's (for those still iterating)
+                    #     and 0's (for those that have completed),  
+                    mask = AVX.bitwise_and(mask, avx1)
+
+                    # Add mask to the iteration counts to keep a count of iterations for each element.
+                    itercounts = AVX.add(itercounts, mask)
+
```

Comment:
Nice touch with the exact derivation :-) Overall, it's very easy to read your code and analysis !